File: LeetCode/23-Merge k Sorted Lists.py
# 主要思路思路是用之前的两列表合并，再加上二分法就可以得到较好的效率
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = cur = ListNode(0)
        while l1 and l2:
            if l1.val < l2.val:
                cur.next = l1
                l1 = l1.next
            else:
                cur.next = l2
                l2 = l2.next
            cur = cur.next
        cur.next = l1 or l2
        return dummy.next
    
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if (len(lists) == 0):
            print ("error")
            return None
        if (len(lists) == 1):
            return lists[0]
        if (len(lists) == 2):
            return self.mergeTwoLists(lists[0], lists[1])
        length = len(lists)
        return self.mergeTwoLists(self.mergeKLists(lists[:length // 2]), self.mergeKLists(lists[length // 2:]))

File: LeetCode/23. Merge k Sorted Lists.cpp
// 可以直接标准的使用堆解决，所有节点扔最小堆，然后弹出拼接就行
// 看了下最快答案，是使用了两两merge，类似二分的方式来做，会快一半左右，可以借鉴下，不过复杂度呀应该都是O(nlogn)，应该只存在常数上的优势
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct cmp{
    bool operator()(ListNode* a, ListNode* b) {
        return a->val > b->val;
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, cmp> pq = priority_queue<ListNode*, vector<ListNode*>, cmp>();
        ListNode* head = NULL;
        ListNode* cur = NULL;
        for (int i = 0; i < lists.size(); ++i) {
            head = lists[i];
            while(head != NULL) {
                pq.push(head);
                head = head->next;
            }
        }
        cur = head = new ListNode();
        while(!pq.empty()) {
            cur->next = pq.top();
            cur = cur->next;
            pq.pop();
        }
        cur->next = NULL;
        return head->next;
    }
};